### PR TITLE
add support for ICMP pcap search

### PR DESCRIPTION
**Pull request by mccanne**
This commit adds support for searching for ICMP flows in pcaps.
It also adds hooks to "pcap slice" and the zqd packet end point
for filtering on icmp.

The commit also cleans up the pcap filtering API and internals by
creating more modular closures for packet filtering  At some
point, we may want to create a Go DSL for creating BPF programs
and port the BPF virtual machine and data flow optimizer to Go.
This would allow for much more flexible filtering, but is not
a priority right now.

### CODE REVIEW

**Comment by philrz** [comment]
I tested out the Brim app at commit `be53d3b` with this branch at commit `0e39433` and it worked well. I was able to open all the ICMP flows I tried in Wireshark, some of which had null `duration` and some >0. 👍

File: pcap/search.go [hunk 1]
```diff
@@ -75,21 +64,74 @@ func (s Search) Span() nano.Span {
...
+
+func genFlowFilter(flow Flow) func(Socket, Socket) bool {
+	return func(s0, s1 Socket) bool {
 		return s0.IP.Equal(flow.S0.IP) && s1.IP.Equal(flow.S1.IP) && s0.Port == flow.S0.Port && s1.Port == flow.S1.Port
 	}
-	return func(tcp bool, src, dst Socket) bool {
-		if tcp != tcp_ {
+}
+
+func genTCPFilter(flow Flow) func(gopacket.Packet) bool {
```
Comment (alfred-landrum) [inline pcap/search.go]: nit: these gen*Filter functions could have return arguments of `PacketFilter`

File: pcap/search.go [hunk 2]
```diff
@@ -75,21 +64,74 @@ func (s Search) Span() nano.Span {
...
+	match := genFlowFilter(flow)
+	return func(packet gopacket.Packet) bool {
+		srcIP, dstIP := matchIP(packet)
+		if srcIP == nil {
```
Comment (alfred-landrum) [inline pcap/search.go]: I spent a few seconds wondering why there's no `dstIP == nil` check here. Would you consider making matchIP return arguments look like `(src net.IP, dst net.IP, ok bool)`?

File: pcap/search.go [hunk 3]
```diff
@@ -75,21 +64,74 @@ func (s Search) Span() nano.Span {
...
-	return s.id
+func genUDPFilter(flow Flow) func(gopacket.Packet) bool {
+	match := genFlowFilter(flow)
+	return func(packet gopacket.Packet) bool {
+		srcIP, dstIP := matchIP(packet)
+		if srcIP == nil {
+			return false
+		}
+		transport := packet.TransportLayer()
+		tcp, ok := transport.(*layers.UDP)
```
Comment (alfred-landrum) [inline pcap/search.go]: Consider using `udp` for the udp layer var instead of tcp.

File: pcap/search.go (hunk 2 above)
Comment (mccanne) [reply pcap/search.go]: ok

File: pcap/search.go (hunk 3 above)
Comment (mccanne) [reply pcap/search.go]: ha!  of course

**Pull request merged**